workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourLibrary.cs

```diff
@@ -0,0 +1,191 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Reflection;
+using Unity.Entities;
+using UnityEngine;
+
+namespace Improbable.Gdk.Core
+{
+    /// <summary>
+    ///     Retrieves Reader and Writer fields from MonoBehaviours and handles injection into them.
+    /// </summary>
+    public class SpatialOSBehaviourLibrary
+    {
+        private readonly Dictionary<Type, Dictionary<uint, FieldInfo>> fieldHandleCache
+            = new Dictionary<Type, Dictionary<uint, FieldInfo>>();
+
+        private readonly Dictionary<Type, List<uint>> componentReaderIdsForBehaviours =
+            new Dictionary<Type, List<uint>>();
+
+        private readonly Dictionary<Type, List<uint>> componentWriterIdsForBehaviours =
+            new Dictionary<Type, List<uint>>();
+
+        private readonly HashSet<Type> invalidMonoBehaviourTypes = new HashSet<Type>();
+
+        private readonly ILogDispatcher logger;
+        private const string LoggerName = "SpatialOSBehaviourLibrary";
+
+        private const string BadRequiredMemberWarning
+            = "[Require] attribute found on member that is not Reader or Writer or incorrectly generated, ignoring this member!";
+
+        private const string MultipleReadersWritersRequiredError
+            = "MonoBehaviour found requesting more than one Reader or Writer for the same component, " +
+            "this is invalid and will not be enabled!";
+
+        private const string MalformedReaderOrWriter
+            = "Reader or Writer found without a Component ID attribute, this is invalid!";
+
+        public SpatialOSBehaviourLibrary(ILogDispatcher logger)
+        {
+            this.logger = logger;
+        }
+
+        public void InjectAllReadersWriters(MonoBehaviour spatialOSBehaviour)
+        {
+            EnsureLoaded(spatialOSBehaviour.GetType());
```

Comment:
`spatialOSBehaviour.GetType()` can be cached since this is used over and over, i.e. `var spatialOSBehaviourType = spatialOSBehaviour.GetType();`.